resources/js/store/store-supporter-tag.tsx

```diff
@@ -0,0 +1,316 @@
+// See the LICENCE file in the repository root for full licence text.
+
+import { UserCard } from 'components/user-card';
+import UserJson from 'interfaces/user-json';
+import { route } from 'laroute';
+import { debounce } from 'lodash';
+import { action, autorun, computed, makeObservable, observable, runInAction } from 'mobx';
+import { disposeOnUnmount, observer } from 'mobx-react';
+import core from 'osu-core-singleton';
+import React from 'react';
+import { onError } from 'utils/ajax';
+import { classWithModifiers } from 'utils/css';
+import { parseJsonNullable, storeJson } from 'utils/json';
+import { trans, transChoice } from 'utils/lang';
+import { toggleCart } from 'utils/store-cart';
+import { currentUrlParams } from 'utils/turbolinks';
+
+const jsonId = 'json-store-supporter-tag';
+
+const maxValue = 52;
+const minValue = 4;
+const resolution = 8;
+
+interface Props {
+  maxMessageLength: number;
+  productId: number;
+}
+
+interface SavedState {
+  giftMessage: string;
+  sliderValue: number;
+  username: string;
+}
+
+interface SliderUIParams {
+  handle?: JQuery | undefined;
+  value?: number | undefined;
+  values?: number[] | undefined;
+}
+
+const monthPresets = [1, 2, 4, 6, 12, 18, 24] as const;
+
+function durationToPrice(duration: number) {
+  switch (true) {
+    case duration >= 12: return Math.ceil(duration / 12.0 * 26);
+    case duration === 10: return 24;
+    case duration === 9: return 22;
+    case duration === 8: return 20;
+    case duration === 6: return 16;
+    case duration === 4: return 12;
+    case duration === 2: return 8;
+    case duration === 1: return 4;
+  }
+}
+
+function sliderValueFrom(value: number) {
+  return value * resolution;
+}
+
+@observer
+export default class StoreSupporterTag extends React.Component<Props> {
+  private readonly debouncedGetUser;
+  private readonly giftMessage: string = '';
+  private readonly giftMessageRef = React.createRef<HTMLTextAreaElement>();
+  private readonly sliderRef = React.createRef<HTMLDivElement>();
+  @observable private sliderValue = sliderValueFrom(minValue);
+  @observable private user: UserJson | null;
+  @observable private username = currentUrlParams().get('target') ?? '';
+  private xhr: JQuery.jqXHR<UserJson> | null = null;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.debouncedGetUser = debounce(this.getUser, 300);
+    document.addEventListener('turbolinks:before-cache', this.handleBeforeCache);
+
+    makeObservable(this);
+
+    const json = parseJsonNullable<SavedState>(jsonId, true);
+    if (json != null) {
+      this.giftMessage = json.giftMessage;
+      this.sliderValue = json.sliderValue;
+      this.username = json.username;
+    }
+
+    if (this.username !== '') {
+      this.user = null;
+      this.debouncedGetUser(this.username);
+    } else {
+      this.user = core.currentUserOrFail;
+    }
+
+    disposeOnUnmount(
+      this,
+      autorun(() => {
+        toggleCart(this.isValidUser);
```

Comment:
this is triggered too early, causing wrong button state if `target` is specified and the username check takes too long or the user doesn't exist

(irrelevant if not doing any workarounds)